addons/portal/models/portal_mixin.py

```diff
@@ -145,3 +154,19 @@ def _sign_token(self, pid):
             "database.secret")
         token = (self.env.cr.dbname, self.access_token, pid)
         return hmac.new(secret.encode('utf-8'), repr(token).encode('utf-8'), hashlib.sha256).hexdigest()
+
+    def _check_token(self, token):
+        """Checks the access token on a given record and returns the portal access and edit rights.
+
+        :param token: (str) The token to check
+        :return: (bool, bool) True if portal access rights granted, True if edit rights granted
+        """
+        self.ensure_one()
+        portal_access = portal_edit = False
+        if token:
+            record_sudo = self.sudo()
+            if record_sudo.access_token and consteq(record_sudo.access_token, token):
+                portal_access = True
+            if record_sudo.edit_token and consteq(record_sudo.edit_token, token):
+                portal_access = portal_edit = True
+        return portal_access, portal_edit
```

Comment:
You could avoid to test portal_access by testing portal_edit first. If portal edit = True, no need to test portal_access.
also, you could use direct assignation

        self.ensure_one()
        record_sudo = self.sudo()
        portal_access = portal_edit = token and record_sudo.edit_token and consteq(record_sudo.edit_token, token)
        if not portal_access:
            portal_access = token and record_sudo.access_token and consteq(record_sudo.access_token, token)
        return portal_access, portal_edit